File: src/main/java/com/sea/dsl/explain/DramaScript.java
package com.sea.dsl.explain;

import groovy.lang.Script;

public abstract class DramaScript extends Script {

    private Context context() {
        return (Context) getBinding().getProperty("context");
    }

    public Variable let(String var) {
        return new Variable(context(), var);
    }

    public DramaObject left(String name) {
        return context().variables.get("leftTable_" + name);
    }

    public DramaObject right(String name) {
        return context().variables.get("rightTable_" + name);
    }

    public Condition when(DramaBoolean dramaBoolean) {
        return new Condition(context(), dramaBoolean);
    }
}

File: src/main/scala/com/sea/dsl/explain/DramaScript.scala
package com.sea.dsl.explain

import groovy.lang.Script
import org.apache.spark.sql.DataFrame
import org.apache.spark.sql.functions.udf

abstract class DramaScript extends Script {


  private def context: DataFrame = {
    getBinding.getProperty("context").asInstanceOf[DataFrame]
  }

  def left(name: String): DramaObject = {
    getBinding.getVariable("leftTable_" + name).asInstanceOf[DramaObject]
  }

  def right(name: String): DramaObject = {
    getBinding.getVariable("rightTable_" + name).asInstanceOf[DramaObject]
  }

  def when(condition: DramaObject): Unit = {
    val brokenRule = Seq("leftTable_p2", "rightTable_p5")
    val explaination = ("y", "y")
    val explainFunc: (Boolean, Map[Seq[String], Seq[(String, String)]]) => Map[Seq[String], Seq[(String, String)]] =
      (b, map) => {
        if (b) {
          if (map.get(brokenRule).isEmpty) {
            map
          } else {
            map + (brokenRule -> (
              Seq(explaination) ++
                map(brokenRule)
              ))
          }
        }
        else
          map
      }
    val explain = udf(explainFunc)
    val temp = context.withColumn("__explain", explain(condition.column, context("__explain")))
    getBinding.setProperty("context", temp)
  }

}
